lib/Backends.hs

```diff
@@ -5,16 +5,51 @@
 module Backends
   ( supportedBackends
   , backendPackedCodec
+  , SomeBackendAPI(..)
   ) where
 
 import Backend (Backend(..), SomeBackend(..))
 import Backend.Vault.Kv (VaultKvBackend)
+import Data.Aeson qualified as A
+import Data.Aeson.Types qualified as A
+import Data.Bifunctor (first)
+import Data.ByteString.Lazy qualified as BSL
 import Data.HashMap.Strict qualified as HS
 import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Text.Encoding qualified as TE
+import Servant.API (FromHttpApiData(..))
 import Toml (TomlCodec)
 import Toml qualified
 import Validation (Validation(Failure))
 
+newtype SomeBackendAPI = SomeBackendAPI SomeBackend
+
+instance A.FromJSON SomeBackendAPI where
+  parseJSON original = A.withObject "SomeBackend" (\obj ->
+    do
+      bType <- lookup "type" obj >>= \case
+        A.String val -> pure val
+        invalid -> keyMismatch "type" invalid
+      case bType of
+        "vault-kv" -> fmap (SomeBackendAPI . SomeBackend) $ parseVaultKv original
+        _ -> fail "Unknown backend type") original
+    where
+      parseVaultKv :: A.Value -> (A.Parser VaultKvBackend)
+      parseVaultKv obj = A.parseJSON obj
+
+      lookup :: Text -> A.Object -> (A.Parser A.Value)
+      lookup key obj = case HS.lookup key obj of
+        Just val -> return val
+        Nothing -> fail $ "No such object with key " ++ (T.unpack key) ++ "."
+
+      keyMismatch :: String -> A.Value -> A.Parser a
+      keyMismatch key inv = A.prependFailure ("Mismatch in " ++ key ++ " ") $ A.typeMismatch "String" inv
+
+instance FromHttpApiData SomeBackendAPI where
+  parseHeader t = (first T.pack) . A.eitherDecode . BSL.fromStrict $ t
```

Comment:
You can use `eitherDecodeStrict` instead of `eitherDecode + fromStrict`.

Also, because we know we'll need to consume the entire `SomeBackendAPI` structure later, we should use the "prime" variant here (`eitherDecodeStrict'` instead of `eitherDecodeStrict`). The parsing functions without `'` build up some thunks during deserialization, whereas the functions with `'` evaluate everything up-front. Therefore:

* functions with `'` are more efficient when you know you're gonna need to consume the entire data structure anyway.
* functions without `'` are more efficient when you might only need to consume a subset of the data structure, and rest is left as unevaluated thunks.

See the [haddocks for `json` vs `json'`](https://hackage.haskell.org/package/aeson-2.1.0.0/docs/Data-Aeson.html#v:json) for details.

```suggestion
  parseHeader = first T.pack . A.eitherDecodeStrict'
```

Reply:
done